File: core/src/com/ariasaproject/advancerofrpg/physics/box2d/PolygonShape.java
package com.ariasaproject.advancerofrpg.physics.box2d;

import com.ariasaproject.advancerofrpg.math.Vector2;

public class PolygonShape extends Shape {
	public PolygonShape() {
		addr = newPolygonShape();
	}

	protected PolygonShape(long addr) {
		this.addr = addr;
	}

	private native long newPolygonShape();

	@Override
	public Type getType() {
		return Type.Polygon;
	}

	public void set(Vector2[] vertices) {
		float[] verts = new float[vertices.length * 2];
		for (int i = 0, j = 0; i < vertices.length * 2; i += 2, j++) {
			verts[i] = vertices[j].x;
			verts[i + 1] = vertices[j].y;
		}
		set(verts, 0, verts.length);
	}

	public void set(float[] vertices) {
		set(vertices, 0, vertices.length);
	}

	public native void set(float[] vertices, int offset, int len); /*
																	 * b2PolygonShape* poly = (b2PolygonShape*)addr; int
																	 * numVertices = len / 2; b2Vec2* verticesOut = new
																	 * b2Vec2[numVertices]; for(int i = 0; i <
																	 * numVertices; i++) { verticesOut[i] =
																	 * b2Vec2(verts[(i<<1) + offset], verts[(i<<1) +
																	 * offset + 1]); } poly->Set(verticesOut,
																	 * numVertices); delete[] verticesOut;
																	 */

	public native void setAsBox(float hx, float hy); /*
														 * b2PolygonShape* poly = (b2PolygonShape*)addr;
														 * poly->SetAsBox(hx, hy);
														 */

	public void setAsBox(float hx, float hy, Vector2 center, float angle) {
		setAsBox(hx, hy, center.x, center.y, angle);
	}

	public native void setAsBox(float hx, float hy, float cx, float cy, float angle); /*
																						 * b2PolygonShape* poly =
																						 * (b2PolygonShape*)addr;
																						 * poly->SetAsBox( hx, hy,
																						 * b2Vec2( centerX, centerY ),
																						 * angle );
																						 */

	public native int getVertexCount();/*
										 * b2PolygonShape* poly = (b2PolygonShape*)addr; return poly->GetVertexCount();
										 */

	private static float[] verts = new float[2];

	public void getVertex(int index, Vector2 vertex) {
		jniGetVertex(addr, index, verts);
		vertex.x = verts[0];
		vertex.y = verts[1];
	}

	private native void jniGetVertex(long addr, int index, float[] verts); /*
																			 * b2PolygonShape* poly =
																			 * (b2PolygonShape*)addr; const b2Vec2 v =
																			 * poly->GetVertex( index ); verts[0] = v.x;
																			 * verts[1] = v.y;
																			 */
}

File: native-core/src/box2d/PolygonShape.cpp
#include "PolygonShape.h"
#include "Shape.h"
#include "Collision/Shapes/b2PolygonShape.h"

PolygonShape_M(jlong, newPolygonShape)(JNIEnv *env, jclass clazz) {
    return (jlong) (new b2PolygonShape());
}

PolygonShape_M(void, set)(JNIEnv *env, jobject obj, jfloatArray vals, jint offset, jint len) {
    jfloat *data = (jfloat *) env->GetPrimitiveArrayCritical(vals, 0);
    int numVertices = len / 2;
    b2Vec2 *verticesOut = new b2Vec2[numVertices];
    for (int i = 0; i < numVertices; i++) {
        verticesOut[i] = b2Vec2(data[(i << 1) + offset], data[(i << 1) + offset + 1]);
    }
    ((b2PolygonShape *) env->GetLongField(obj, shapePtr))->Set(verticesOut, numVertices);
    delete[] verticesOut;
    env->ReleasePrimitiveArrayCritical(vals, data, 0);
}

PolygonShape_M(void, setAsBox___3FF)(JNIEnv *env, jobject obj, jfloat hx, jfloat hy) {
    ((b2PolygonShape *) env->GetLongField(obj, shapePtr))->SetAsBox(hx, hy);
}

PolygonShape_M(void, setAsBox___3FFFFF)(JNIEnv *env, jobject obj, jfloat hx, jfloat hy, jfloat cx,
                                        jfloat cy, jfloat angle) {
    ((b2PolygonShape *) env->GetLongField(obj, shapePtr))->SetAsBox(hx, hy, b2Vec2(cx, cy), angle);
}

PolygonShape_M(jint, getVertexCount)(JNIEnv *env, jobject obj) {
    return ((b2PolygonShape *) env->GetLongField(obj, shapePtr))->GetVertexCount();
}

PolygonShape_M(void, jniGetVertex)(JNIEnv *env, jobject obj, jint index, jfloatArray vals) {
    jfloat *data = (jfloat *) env->GetPrimitiveArrayCritical(vals, 0);
    const b2Vec2 v = ((b2PolygonShape *) env->GetLongField(obj, shapePtr))->GetVertex(index);
    data[0] = v.x;
    data[1] = v.y;
    env->ReleasePrimitiveArrayCritical(vals, data, 0);
}


